File: scripts/api_test.py
import requests
import json
import os
import logging

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

class TradingApiReport:
    def __init__(self, username, auth_token):
        self.username = username
        self.auth_token = auth_token
        self.url = 'https://www.trade-copier.com/webservice/v4/reporting/getReporting.php'

    def get_reports(self, year):
        reports = []
        headers = {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Auth-Username': self.username,
            'Auth-Token': self.auth_token,
        }

        # Iterate through each month to get reports
        for month in range(1, 13):  # Months are 1-12
            response = requests.post(self.url, headers=headers, data={'month': month, 'year': year})
            
            # Check if the request was successful
            if response.status_code == 200:
                data = response.json()
                
                if 'reporting' in data:
                    for report in data['reporting']:
                        reports.append(report)
                else:
                    logging.error('Invalid JSON or missing reporting key')
            else:
                logging.error(f"HTTP Error: {response.status_code} - {response.text}")

        self.save_reports_to_json(reports)
        return reports

    def save_reports_to_json(self, reports, filename='leaderboard.json'):
        storage_path = 'storage/app'
        os.makedirs(storage_path, exist_ok=True)
        file_path = os.path.join(storage_path, filename)

        try:
            with open(file_path, 'w') as json_file:
                json.dump(reports, json_file, indent=4)
            logging.info(f"Reports data saved to {file_path}")
        except Exception as e:
            logging.error(f"Failed to save reports data: {e}")

username = 'jamespereiraofficial'
auth_token = 'jM3NTc1N2YxNDFmODIwZjQ3NTg3OWR'
year = 2024

trading_api_report = TradingApiReport(username, auth_token)
reports = trading_api_report.get_reports(year)

#print("Reports:", reports)

File: scripts/api_test.js
const https = require('https');
const fs = require('fs');
const path = require('path');

// Simple logging functions
const logInfo = (message) => console.info(`${new Date().toISOString()} - INFO - ${message}`);
const logError = (message) => console.error(`${new Date().toISOString()} - ERROR - ${message}`);

class TradingApiReport {
  constructor(username, authToken) {
    this.username = username;
    this.authToken = authToken;
    this.url = 'https://www.trade-copier.com/webservice/v4/reporting/getReporting.php';
  }

  // Method to make HTTP POST requests without third-party libraries
  makePostRequest(url, data, headers, callback) {
    const urlObject = new URL(url);
    const options = {
      hostname: urlObject.hostname,
      path: urlObject.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        ...headers
      }
    };

    const req = https.request(options, (res) => {
      let body = '';

      res.on('data', (chunk) => {
        body += chunk;
      });

      res.on('end', () => {
        callback(null, body, res);
      });
    });

    req.on('error', (error) => {
      callback(error);
    });

    req.write(data);
    req.end();
  }

  async getReports(year) {
    const reports = [];
    const headers = {
      'Auth-Username': this.username,
      'Auth-Token': this.authToken
    };

    for (let month = 1; month <= 12; month++) {
      try {
        const data = `month=${month}&year=${year}`;
        const response = await new Promise((resolve, reject) => {
          this.makePostRequest(this.url, data, headers, (err, body, res) => {
            if (err) return reject(err);
            resolve({ body, statusCode: res.statusCode });
          });
        });

        if (response.statusCode === 200) {
          const data = JSON.parse(response.body);

          if (data && data.reporting) {
            reports.push(...data.reporting);
          } else {
            logError('Invalid JSON or missing reporting key');
          }
        } else {
          logError(`HTTP Error: ${response.statusCode} - ${response.body}`);
        }
      } catch (error) {
        logError(`Error fetching data for month ${month}: ${error.message}`);
      }
    }

    this.saveReportsToJson(reports);
    return reports;
  }

  saveReportsToJson(reports, filename = 'leaderboard.json') {
    const storagePath = path.join(__dirname, 'storage', 'app');
    const filePath = path.join(storagePath, filename);

    fs.mkdir(storagePath, { recursive: true }, (err) => {
      if (err) {
        return logError(`Failed to create directory: ${err.message}`);
      }

      fs.writeFile(filePath, JSON.stringify(reports, null, 4), (err) => {
        if (err) {
          return logError(`Failed to save reports data: ${err.message}`);
        }
        logInfo(`Reports data saved to ${filePath}`);
      });
    });
  }
}

// Usage
const username = 'jamespereiraofficial';
const authToken = 'jM3NTc1N2YxNDFmODIwZjQ3NTg3OWR';
const year = 2024;

const tradingApiReport = new TradingApiReport(username, authToken);
tradingApiReport.getReports(year).then((reports) => {
  console.log('Reports:', reports);
}).catch(error => {
  logError(`Failed to fetch reports: ${error.message}`);
});
